mirgecom/logging_quantities.py

```diff
@@ -0,0 +1,319 @@
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+__doc__ = """
+.. autoclass:: StateConsumer
+.. autoclass:: DiscretizationBasedQuantity
+.. autoclass:: KernelProfile
+.. autoclass:: PythonMemoryUsage
+.. autofunction:: initialize_logmgr
+.. autofunction:: logmgr_add_device_name
+.. autofunction:: logmgr_add_default_discretization_quantities
+.. autofunction:: add_package_versions
+.. autofunction:: set_sim_state
+"""
+
+from logpyle import (LogQuantity, LogManager, MultiLogQuantity, add_run_info,
+    add_general_quantities, add_simulation_quantities)
+from meshmode.array_context import PyOpenCLArrayContext
+from meshmode.discretization import Discretization
+import pyopencl as cl
+
+from typing import Optional, Callable
+import numpy as np
+
+
+def initialize_logmgr(enable_logmgr: bool, enable_profiling: bool,
+                      filename: str = None, mode: str = "wu",
+                      mpi_comm=None) -> LogManager:
+    """Create and initialize a mirgecom-specific :class:`logpyle.LogManager`."""
+    if not enable_logmgr:
+        return None
+
+    logmgr = LogManager(filename=filename, mode=mode, mpi_comm=mpi_comm)
+
+    add_run_info(logmgr)
+    add_package_versions(logmgr)
+    add_general_quantities(logmgr)
+    add_simulation_quantities(logmgr)
+
+    try:
+        logmgr.add_quantity(PythonMemoryUsage())
+    except ImportError:
+        from warnings import warn
+        warn("psutil module not found, not tracking memory consumption."
+             "Install it with 'pip install psutil'")
+
+    return logmgr
+
+
+def logmgr_add_device_name(logmgr: LogManager, queue: cl.CommandQueue):
+    """Add the OpenCL device name to the log."""
+    logmgr.set_constant("cl_device_name", str(queue.device))
+
+
+def logmgr_add_default_discretization_quantities(logmgr: LogManager, discr, dim,
+      extract_vars_for_logging, units_for_logging):
+    """Add default discretization quantities to the logmgr."""
+    for quantity in ["pressure", "temperature"]:
+        for op in ["min", "max", "norm"]:
+            logmgr.add_quantity(DiscretizationBasedQuantity(
+                discr, quantity, op, extract_vars_for_logging, units_for_logging))
+    for quantity in ["mass", "energy"]:
+        for op in ["min", "max", "norm"]:
+            logmgr.add_quantity(DiscretizationBasedQuantity(
+                discr, quantity, op, extract_vars_for_logging, units_for_logging))
+
+    for dim in range(dim):
+        for op in ["min", "max", "norm"]:
+            logmgr.add_quantity(DiscretizationBasedQuantity(
+                discr, "momentum", op, extract_vars_for_logging, units_for_logging,
+                axis=dim))
```

Comment:
Computing all those will be quite expensive, since it's lots of reductions (the GPU kind). I'm not sure we should do that by default. Maybe allow the user to specify a list of strings for what they actually want, and use a more parsimonious default.

Reply:
Note that these aren't activated by default. Maybe just rename the function?

Reply:
Yeah, that would help.